src/ngx_http_lua_timer.c

```diff
@@ -129,6 +270,78 @@ ngx_http_lua_ngx_timer_every(lua_State *L)
 }
 
 
+static int 
+ngx_http_lua_ngx_timer_cancel(lua_State *L) 
+{
+    int                             nargs;
+    ngx_event_t                     *ev;
+    ngx_http_request_t              *r;
+    ngx_http_lua_main_conf_t        *lmcf;
+    ngx_http_lua_timer_ctx_t        tctx;
+
+    nargs = lua_gettop(L);
+
+    if (nargs < 1) {
+        return luaL_error(L, "expecting at least 1 arguments but got %d",
+                          nargs);
+    }
+
+    unsigned ref = (unsigned)luaL_checknumber(L, 1);
+
+    if (ref >= reftable->max_refs || !bit_check(ref))  {
+        lua_pushnil(L);
+        lua_pushliteral(L, "timer does not exist");
+        return 2;
+    }
+    
+    r = ngx_http_lua_get_req(L);
+    if (r == NULL) {
+        return luaL_error(L, "no request");
+    }
+    
+    ev = reftable->ref[ref];
+
+    ngx_memcpy(&tctx, ev->data, sizeof(ngx_http_lua_timer_ctx_t));
+    
+    if (tctx.delay) {
+        ngx_del_timer(ev);
+    } else {
+
+#ifdef HAVE_POSTED_DELAYED_EVENTS_PATCH
+        ngx_delete_posted_event(ev);
+#else
+        ngx_del_timer(ev);
+#endif
+
+    }
+    
+    ngx_free(ev);
+
+    ngx_http_lua_assert(tctx.co_ref && tctx.co);
+
+    lmcf = tctx.lmcf;
+    lmcf->pending_timers--;
+
+
+    ngx_http_lua_free_thread(r, L, tctx.co_ref, tctx.co, lmcf);
+
+    if (tctx.vm_state != NULL) {
+        ngx_http_lua_cleanup_vm(tctx.vm_state);
+    }
+
+    if (tctx.pool) {
+        ngx_destroy_pool(tctx.pool);
+    }
+
+    bit_set(ref, 0);
+
+    if (ref < reftable->next_ref) reftable->next_ref = ref;
```

Comment:
Can we wrap the bit set 1/0 to two helper functions?

Reply:
Sure, use bit_set and bit_clean instead.